src/goto-programs/remove_exceptions.cpp

```diff
@@ -0,0 +1,554 @@
+  {
+    for(std::size_t j=stack_catch[i].size(); j-->0;)
+    {
+      goto_programt::targett new_state_pc=
+        stack_catch[i][j].second;
+
+      // find handler
+      goto_programt::targett t_exc=goto_program.insert_after(instr_it);
+      t_exc->make_goto(new_state_pc);
+      t_exc->source_location=instr_it->source_location;
+      t_exc->function=instr_it->function;
+
+      // use instanceof to check that this is the correct handler
+      symbol_typet type(stack_catch[i][j].first);
+      type_exprt expr(type);
+
+      binary_predicate_exprt check(exc_thrown, ID_java_instanceof, expr);
+      t_exc->guard=check;
+    }
+  }
+
+  // add dead instructions
+  for(const auto &local : locals)
+  {
+    goto_programt::targett t_dead=goto_program.insert_after(instr_it);
+    t_dead->make_dead();
+    t_dead->code=code_deadt(local);
+    t_dead->source_location=instr_it->source_location;
+    t_dead->function=instr_it->function;
+  }
+
+  // replace "throw x;" by "f#exception_value=x;"
+  exprt exc_expr=instr_it->code;
+  while(exc_expr.id()!=ID_symbol && exc_expr.has_operands())
+    exc_expr=exc_expr.op0();
+
+  // add the assignment with the appropriate cast
+  code_assignt assignment(typecast_exprt(exc_thrown, exc_expr.type()),
+                          exc_expr);
+  // now turn the `throw' into `assignment'
+  instr_it->type=ASSIGN;
+  instr_it->code=assignment;
+}
+
+/*******************************************************************\
+
+Function: remove_exceptionst::instrument_function_call
+
+Inputs:
+
+Outputs:
+
+Purpose: instruments each function call that may escape exceptions
+         with conditional GOTOS to the corresponding exception handlers
+
+\*******************************************************************/
+
+void remove_exceptionst::instrument_function_call(
+  const goto_functionst::function_mapt::iterator &func_it,
+  const goto_programt::instructionst::iterator &instr_it,
+  const stack_catcht &stack_catch,
+  std::vector<exprt> &locals)
+{
+  assert(instr_it->type==FUNCTION_CALL);
+
+  goto_programt &goto_program=func_it->second.body;
+  const irep_idt &function_id=func_it->first;
+
+  // save the address of the next instruction
+  goto_programt::instructionst::iterator next_it=instr_it;
+  next_it++;
+
+  code_function_callt &function_call=to_code_function_call(instr_it->code);
+  assert(function_call.function().id()==ID_symbol);
+  const irep_idt &callee_id=
+    to_symbol_expr(function_call.function()).get_identifier();
+
+  if(symbol_table.has_symbol(id2string(callee_id)+EXC_SUFFIX))
+  {
+    // we may have an escaping exception
+    const symbolt &callee_exc_symbol=
+      symbol_table.lookup(id2string(callee_id)+EXC_SUFFIX);
+    symbol_exprt callee_exc=callee_exc_symbol.symbol_expr();
+
+    // // find the end of the function
+    goto_programt::targett end_function;
+    for(end_function=instr_it;
+        !end_function->is_end_function();
+        end_function++) {}
```

Comment:
See above; it might actually be worthwhile adding a method `get_end_function` to `goto_program_templatet`